File: wintest/ui/web/frontend/src/docs/click.ts
import type { StepDoc } from './types';

export const clickDoc: StepDoc = {
  name: 'click',
  title: 'Click',
  summary: 'Click at captured coordinates, or on a UI element identified by AI.',
  description:
    'Performs a click at a screen location. The click_type parameter selects single left, double, right, or middle click. There are two modes: coordinate-based (the default, set up by the Test Builder) uses pixel-exact coordinates recorded when you clicked on the screenshot; AI-based uses a vision model to locate an element described in plain language. Coordinate mode is faster and deterministic; AI mode is useful when a layout may shift between runs.',
  parameters: [
    {
      name: 'click_type',
      type: 'string',
      required: false,
      description:
        'One of "click" (left, default), "double_click", "right_click", or "middle_click".',
    },
    {
      name: 'click_x',
      type: 'number',
      required: false,
      description:
        'Normalized horizontal click coordinate (0.0–1.0). Used together with click_y for coordinate-based clicking. Typically recorded by the Test Builder.',
    },
    {
      name: 'click_y',
      type: 'number',
      required: false,
      description:
        'Normalized vertical click coordinate (0.0–1.0). Used together with click_x for coordinate-based clicking.',
    },
    {
      name: 'target',
      type: 'string',
      required: false,
      description:
        'Natural language description of the UI element to click (AI mode). Only used if click_x and click_y are not set. Example: "Save button", "File menu".',
    },
  ],
  example:
    '- action: click\n  click_type: double_click\n  click_x: 0.35\n  click_y: 0.62\n  description: "Open the report file"',
};

File: wintest/steps/click.py
"""Click step -- click on a UI element or at explicit coordinates.

Supports click types: click (left), double_click, right_click, middle_click.
"""

from ._base import StepDefinition, FieldDef


def validate(step, step_num):
    if not step.target and step.click_x is None:
        return [f"Step {step_num}: 'click' requires a 'target' field or click_x/click_y coordinates"]
    if step.click_type not in ("click", "double_click", "right_click", "middle_click"):
        return [f"Step {step_num}: invalid click_type '{step.click_type}'"]
    return []


def execute(step, agent):
    click_type = step.click_type or "click"
    if step.click_x is not None and step.click_y is not None:
        return agent.click_at(step, click_type=click_type)
    return agent.find_and_click(step, click_type=click_type)


definition = StepDefinition(
    name="click",
    description="Click on a UI element or at coordinates (left, double, right, or middle click)",
    fields=[
        FieldDef("target", "string"),
        FieldDef("click_x", "number"),
        FieldDef("click_y", "number"),
        FieldDef("click_type", "string"),
    ],
    validate=validate,
    execute=execute,
)
